src/main/scala/chisel3/experimental/util/algorithm/Bitwise.scala

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.experimental.util.algorithm
+
+import chisel3._
+
+/** Map each bits to logical or of itself and all bits less siginificant than it.
+  * @example {{{
+  * LSBOr("b00001000".U) // Returns "b11111000".U
+  * LSBOr("b00010100".U) // Returns "b11111100".U
+  * LSBOr("b00000000".U) // Returns "b00000000".U
+  * }}}
+  * This circuit seems to be high fan out, but synthesis tool should handle this.
+  */
+object LSBOr {
+  def apply(data: UInt): UInt = {
+    val width = data.widthOption match {
+      case Some(w) => w
+      case None    => throw new IllegalArgumentException("Cannot call LSBOr on data with unknown width.")
+    }
+    VecInit(Seq.tabulate(width) { i: Int =>
+      VecInit(data.asBools().dropRight(width - i - 1)).asUInt().orR()
+    }).asUInt()
+  }
```

Comment:
Things to consider about the implementation. @terpstra noted that the version in rocket-chip is optimized for simulation speed (minimizing the amount of generated Verilog). This is obviously very sensitive to both the number of FIRRTL nodes generated and to the optimizations in the FIRRTL implementation (maybe CIRCT can/will do something different here).

Another thing to consider (h/t @sequencer) is how well things will constant propagate if you were to call this then just index the MSB. It would be ideal if that were optimized but is not clear if that's reasonable to expect a FIRRTL compiler to do.

Reply:
There was a common usage: "find first one" using this API. However can simulator detect this pattern or not is unclear, I think optimization pass to detect this common circuit should happen in ventilator or VCS passes.
Node for this compact mode is about N(indexing) + N(orR) ~2N, generally amd64 backend can leverage more CISC benefits here.
Essentially, this is the prefix tree for each bit, a prefix graph for entire circuit. node for prefix tree depends on the the tree implementation, but mostly higher than 2N, if my calculation is correct.

Maybe implementing a prefix tree in Chisel sounds a final idea? I'm not so sure about this. How to design API with a prefix graph.(maybe an implicit class? not so sure)

Reply:
What would we loose by just copying over the rocket-chip solution? That seems to be by far the simplest solution. Are there actually noticable QoR differences?

Reply:
From dev today, I think we should stick with the rocket-chip implementation (even if we don't expose the `width` arguments). Looking at how many nodes it creates, this implementation is `O(n^2)` vs. `O(log(n))` for the rocket-chip implementation.

Reply:
The implementation has been changed to stay in line with rocket-chip in e6fab506430c1bce4ce5e0de64cba8e08658d3f6.